### PR TITLE
Upgrade gradle to 5.3.1

### PR DESCRIPTION
**Pull request by timja**
I'm looking at java 11 compatibility, saw this version is quite old which makes it harder to compile with newer java gradle 5.x is required for full compatibility

For significant contributions please make sure you have completed the following items:

- [na] Design discussion issue #
- [na] Changes in public surface reviewed
- [x] CHANGELOG.md updated
I would suggest using something like https://github.com/apps/release-drafter to maintain the CHANGELOG, much easier...

### CODE REVIEW

File: distributions/build.gradle [hunk 1]
```diff
@@ -73,4 +73,4 @@ gradle.projectsEvaluated {
 }
 
 // Building the shadow (fat) jar after compiling sources.
-tasks.assemble.doLast { tasks.shadowJar.execute() }
+assemble.dependsOn(shadowJar)
```
Comment (dhaval24) [inline distributions/build.gradle]: It's a good practice to have end of line character. I would recommend adding it here.

**Comment by timja** [comment]
The smoke test build is a private project and not using the Azure Pipelines app so output on the GitHub check 😢 

What'd it fail on?

**Comment by dhaval24** [comment]
@littleaj would you help please trigger CI and smoke tests for this PR?

**Comment by dhaval24** [comment]
@timja I tried to build you branch locally on windows machine and I am getting the following errors. 

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':core:linkWindowsX86SharedLibrary'.
> A build operation failed.
      Linker failed while linking applicationinsights-core-native-win32dbg.dll.
  See the complete log at: file:///E:/ApplicationInsights-Java/core/build/tmp/linkWindowsX86SharedLibrary/output.txt
   > Linker failed while linking applicationinsights-core-native-win32dbg.dll.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':core:linkWindowsX64SharedLibrary'.
> A build operation failed.
      Linker failed while linking applicationinsights-core-native-win64dbg.dll.
  See the complete log at: file:///E:/ApplicationInsights-Java/core/build/tmp/linkWindowsX64SharedLibrary/output.txt
   > Linker failed while linking applicationinsights-core-native-win64dbg.dll.

Also when I try to build it without native libraries I still get the following:

> Task :collectd:assemble FAILED

FAILURE: Build failed with an exception.

* Where:
Build file 'E:\ApplicationInsights-Java\collectd\build.gradle' line: 69

* What went wrong:
Execution failed for task ':collectd:assemble'.
> Could not find method execute() for arguments [] on task ':collectd:shadowJar' of type com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.3.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 56s
71 actionable tasks: 57 executed, 14 up-to-date

I am building this on Azul OpenJDK 8.

**Comment by timja** [comment]
Not surprised that didn't run on my machine `linkWindowsX64SharedLibrary'`
I'll spin up a windows VM and take a look

**Comment by timja** [comment]
@dhaval24 are you able to try again?

I haven't managed to build this on windows yet currently failing with:
```
> Task :core:compileWindowsX86SharedLibraryWindowsCpp FAILED
In file included from C:\Users\timja\projects\ApplicationInsights-Java\core\src\windows\cpp\CppPerfCounters.cpp:24:
C:\Users\timja\projects\ApplicationInsights-Java\core\src\windows\cpp\CppPerfCounters.h:22:10: fatal error: jni.h: No such file or directory
 #include <jni.h>
```

Will try and figure out why its not finding it...

**Comment by timja** [comment]
Any tips for setting up the native.gradle bit to work properly?

Here are my env vars:
```
Name                           Value
----                           -----
ALLUSERSPROFILE                C:\ProgramData
APPDATA                        C:\Users\timja\AppData\Roaming
ChocolateyInstall              C:\ProgramData\chocolatey
ChocolateyLastPathUpdate       132000661690811205
CLASSPATH                      .;
CLIENTNAME                     Frankel.local
CommonProgramFiles             C:\Program Files\Common Files
CommonProgramFiles(x86)        C:\Program Files (x86)\Common Files
CommonProgramW6432             C:\Program Files\Common Files
COMPUTERNAME                   timj-temp-win10
ComSpec                        C:\windows\system32\cmd.exe
DriverData                     C:\Windows\System32\Drivers\DriverData
HOMEDRIVE                      C:
HOMEPATH                       \Users\timja
JAVA_HOME                      C:\Program Files\Java\zulu8.38.0.13-ca-jdk8.0.212-win_x64
JAVA_JRE_7                     C:\Program Files\Java\jre7
LOCALAPPDATA                   C:\Users\timja\AppData\Local
LOGONSERVER                    \\timj-temp-win10
NUMBER_OF_PROCESSORS           2
OneDrive                       C:\Users\timja\OneDrive
OS                             Windows_NT
Path                           C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\OpenSSH\;C:\ProgramData\chocolatey\bin;C:\Progra...
PATHEXT                        .COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
PROCESSOR_ARCHITECTURE         AMD64
PROCESSOR_IDENTIFIER           Intel64 Family 6 Model 79 Stepping 1, GenuineIntel
PROCESSOR_LEVEL                6
PROCESSOR_REVISION             4f01
ProgramData                    C:\ProgramData
ProgramFiles                   C:\Program Files
ProgramFiles(x86)              C:\Program Files (x86)
ProgramW6432                   C:\Program Files
PSModulePath                   C:\Users\timja\Documents\WindowsPowerShell\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\windows\system32\WindowsPowerShell\v1.0\Modules
PUBLIC                         C:\Users\Public
SESSIONNAME                    RDP-Tcp#1
SystemDrive                    C:
SystemRoot                     C:\windows
TEMP                           C:\Users\timja\AppData\Local\Temp
TMP                            C:\Users\timja\AppData\Local\Temp
USERDOMAIN                     timj-temp-win10
USERDOMAIN_ROAMINGPROFILE      timj-temp-win10
USERNAME                       timja
USERPROFILE                    C:\Users\timja
windir                         C:\windows
```

**Comment by timja** [comment]
Build and smoke tests pass for me with skipWinNative=true set on windows
I'll fiddle for a bit more to try get the native bits working

**Comment by dhaval24** [comment]
@timja you need visual c++ 2013 and above to build native code. I am going to give a shot with your latest commits. Thanks for being patient here.

**Comment by dhaval24** [comment]
@timja I tried to rebuild with the latest changes, and yes it builds fine when skipping native portion but it seems that I get similar exception when building native as well;

linking applicationinsights-core-native-win64dbg.dll failed.
CppPerfCounters.obj : MSIL module encountered; incremental linking is disabled for MSIL; performing full link
LINK : fatal error LNK1104: cannot open file 'MSCOREE.lib'

It appears that between gradle 4.1 and 5.3 there were some changes that were introduced in the way gradle handles building native modules. I am not aware of them though. Do you have a clue?

**Comment by timja** [comment]
I’ll try build on master on Monday / Tuesday to get a working version and then try see what’s different

**Comment by timja** [comment]
Tried on 4.1 and got a similar error on master:
```
LINK : fatal error LNK1104: cannot open file 'MSCOREE.lib'
```

**Comment by littleaj** [comment]
@timja, I had the same issue when setting up the build server. You can try searching for it in `\Program Files (x86)\Microsoft SDKs\` or `\Windows` and copy it into the expected directory; Unfortunately, it's been so long I don't recall where I found it or where the build expects it to be. 🤔 If you want to dig into it, I think I found the expected location either by using `--info` or `--debug`; or building `core/src/windows/cpp/CppPerfCounters.vcxproj` alone in Visual Studio. There is some error which shows the full path.

In any case, I think it's safe to ignore. Use the `-DskipWinNative=true` option and address any other issues that come up. I'll try it on the build server and make sure it's working before merging.

Thanks for all your help with this!

**Comment by timja** [comment]
From my side it builds fine with skip native

I can't get native to build on master or in this branch currently

**Comment by littleaj** [comment]
@timja, ok great. Is it the same error building on master?

**Comment by timja** [comment]
Yes
`LINK : fatal error LNK1104: cannot open file 'MSCOREE.lib'1

**Comment by timja** [comment]
I've managed to build it on master now
Required installing Visual Studio 2013 express (and signing up for a visual studio subscription to download that now...)

and running it from VS2015 x86 Native Tools Command Prompt

Fails on my branch with: 
```
> Task :core:linkWindowsX86SharedLibrary FAILED
CppPerfCounters.obj : MSIL module encountered; incremental linking is disabled for MSIL; performing full link
LINK : fatal error LNK1104: cannot open file 'MSCOREE.lib'
```

**Comment by littleaj** [comment]
@timja, I'm seeing the same error on the build server. It seems something has changed in the native plugin.

**Comment by timja** [comment]
Can you run it with `--info` on the build server?
I've found the difference for me locally

Its picking up the 2019 toolchain (140) instead of the 2013 one

this branch:
```
See file:///C:/Users/timja/projects/ApplicationInsights-Java/core/build/tmp/linkWindowsX86SharedLibrary/output.txt for all output for linkWindowsX86SharedLibrary.
Starting process 'command 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.20.27508\bin\HostX64\x86\link.exe''. Working directory: C:\Users\timja\projects\ApplicationInsights-Java\core Command: C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.20.27508\bin\HostX64\x86\link.exe @C:\Users\timja\projects\ApplicationInsights-Java\core\build\tmp\linkWindowsX86SharedLibrary\options.txt
Successfully started process 'command 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.20.27508\bin\HostX64\x86\link.exe''
CppPerfCounters.obj : MSIL module encountered; incremental linking is disabled for MSIL; performing full link
LINK : fatal error LNK1104: cannot open file 'MSCOREE.lib'
```

master
```
See file:///C:/Users/timja/projects/ApplicationInsights-Java/core/build/tmp/linkWindowsX86SharedLibrary/output.txt for all output for linkWindowsX86SharedLibrary.
Starting process 'command 'C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64_x86\link.exe''. Working directory: C:\Users\timja\projects\ApplicationInsights-Java\core\build\libs\windows\shared\x86 Command: C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64_x86\link.exe @C:\Users\timja\projects\ApplicationInsights-Java\core\build\tmp\linkWindowsX86SharedLibrary\options.txt
Successfully started process 'command 'C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64_x86\link.exe''
   Creating library C:\Users\timja\projects\ApplicationInsights-Java\core\build\libs\windows\shared\x86\applicationinsights-core-native-win32dbg.lib and object C:\Users\timja\projects\ApplicationInsights-Java\core\build\libs\windows\shared\x86\applicationinsights-core-native-win32dbg.exp
```

**Comment by timja** [comment]
@littleaj It compiled for me...

If I add this to the tool chain block of native.gradle

```
C:\Users\timja\projects\ApplicationInsights-Java>git diff
diff --git a/core/native.gradle b/core/native.gradle
index 2a0b9eb6..e185a455 100644
--- a/core/native.gradle
+++ b/core/native.gradle
@@ -98,6 +98,8 @@ model {
                     // Visual Studio Compiler Options: https://msdn.microsoft.com/en-us/library/fwkeyyhe.aspx
                     // Visual Studio Linker Options:   https://msdn.microsoft.com/en-us/library/y0zzbyt4.aspx

+                                       toolChain.installDir = "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\bin\\amd64_x86\\link.exe"
+
                     // Searches a directory for include files.
                     cppCompiler.args "/I$javaIncludeDir"
                     cppCompiler.args "/I$javaIncludeDir\\win32"
```

Thoughts?

**Comment by timja** [comment]
Ugh my mistake, I was on master when I built it before
That line didn't fix it

But the issue is that for me gradle is picking up a newer toolchain than it was before

**Comment by timja** [comment]
Is there any reason to stay on something so old? could it be upgraded to be made compatible?

**Comment by timja** [comment]
I suspect it was this PR that's caused the issue:
https://github.com/gradle/gradle/pull/3280

gradle 4.4

**Comment by timja** [comment]
Its using `"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe"`

to locate it

On my machine that shows Visual Studio Community 2019

**Comment by littleaj** [comment]
@timja, I got it working locally by adding this inside `binaries.all {`:
```groovy
                    if (targetPlatform == platforms.x86) {
                        linker.args "/LIBPATH:C:\\Program Files (x86)\\Windows Kits\\NETFXSDK\\4.6.1\\Lib\\um\\x86"
                    } else if (targetPlatform == platforms.x64) {
                        linker.args "/LIBPATH:C:\\Program Files (x86)\\Windows Kits\\NETFXSDK\\4.6.1\\Lib\\um\\x64"
                    }
```

I'd prefer to add the equivalent option to the vcxproj file. I think then it will detect the location automatically. But that seems to be all that's missing: a lib path.

This was using the VS 2017 tools.

**Comment by littleaj** [comment]
@timja, Unless you've found a way to get vcxproj to pick up the correct lib path, I have a workaround in the `buildme/gradle-upgrade` branch. You can cherry-pick or reset that commit and push it here.

https://github.com/Microsoft/ApplicationInsights-Java/commit/23f883e4ed2f397ec9ccdbfba966346fe414c64a

It works locally for me and on our build server, but with two different paths.
Locally, I have `WIN_SDK_LIB_PATH=C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\Lib`. You should find the mscoree.lib somewhere under `C:\Program Files (x86)\Windows Kits`. Since you mentioned the latest version of VS, I expect it will be in under NETFXSDK.

**Comment by timja** [comment]
Works on my VM with the same ENV var you posted, thanks for your fix!

Anything holding up merging of this?

**Comment by littleaj** [comment]
@timja, There is one thing breaking the smoketests, but I've fixed it and will merge it in a different PR. I'll link to this one if you are interested.

But the main build works. Thanks again for this contribution!

**Pull request merged**